src/commons.jl

```diff
@@ -0,0 +1,199 @@
+	if pos == 0
+		return convert(HKEY, 0)
+	else
+		root_key = uppercase(key[1:(pos-1)])
+	end
+	
+	if root_key == "HKEY_CLASSES_ROOT" || root_key == "HKCR"
+		return ROOT_HKEY.HKEY_CLASSES_ROOT
+	elseif root_key == "HKEY_CURRENT_CONFIG" || root_key == "HKCC"
+		return ROOT_HKEY.HKEY_CURRENT_CONFIG
+	elseif root_key == "HKEY_CURRENT_USER" || root_key == "HKCU"
+		return ROOT_HKEY.HKEY_CURRENT_USER
+	elseif root_key == "HKEY_LOCAL_MACHINE" || root_key == "HKLM"
+		return ROOT_HKEY.HKEY_LOCAL_MACHINE
+	elseif root_key == "HKEY_USERS" || root_key == "HKU"
+		return ROOT_HKEY.HKEY_USERS
+	elseif root_key == "HKEY_PERFORMANCE_DATA"
+		return ROOT_HKEY.HKEY_PERFORMANCE_DATA
+	elseif root_key == "HKEY_PERFORMANCE_TEXT"
+		return ROOT_HKEY.HKEY_PERFORMANCE_TEXT
+	elseif root_key == "HKEY_PERFORMANCE_NLSTEXT"
+		return ROOT_HKEY.HKEY_PERFORMANCE_NLSTEXT
+	elseif root_key == "HKEY_CURRENT_CONFIG" || root_key == "HKCG"
+		return ROOT_HKEY.HKEY_CURRENT_CONFIG
+	elseif root_key == "HKEY_DYN_DATA" || root_key == "HKDD"
+		return ROOT_HKEY.HKEY_DYN_DATA
+	else
+		return convert(HKEY, 0)
+	end
+end
+
+function getsubkey(key::AbstractString)
+	local pos = search(key,'\\')
+	if pos == 0
+		return ""
+	else
+		return key[(pos+1):end]
+	end
+end
+
+function gettype(typ::UInt32)
+	if typ == REG_TYPE.REG_SZ || typ == REG_TYPE.REG_EXPAND_SZ || typ == REG_TYPE.REG_LINK
+		return AbstractString
+	elseif typ == REG_TYPE.REG_MULTI_SZ
+		return Vector{AbstractString}
+	elseif typ == REG_TYPE.REG_DWORD_LITTLE_ENDIAN || typ == REG_TYPE.REG_DWORD_BIG_ENDIAN
+		return UInt32
+	elseif typ == REG_TYPE.REG_QWORD
+		return UInt64
+	elseif typ == REG_TYPE.REG_BINARY
+		return Vector{UInt8}
+	else
+		return 0
+	end
+end
+
+function regopenkeyex(root::HKEY, subkey::AbstractString, sam::REGSAM)
+	local hk = Ref{HKEY}(0)
+	
+	#UNICODE version; xxxExA is ASCII version, but LPCTRSTR = LPCSTR
+	res = ccall((:RegOpenKeyExW, advapi), stdcall,
+	  LONG,
+	  (HKEY, LPCWSTR, DWORD, REGSAM, PHKEY),
+	  root, transcode(Cwchar_t, subkey), 0, sam, hk )
+	
+	if res == ERROR_SUCCESS
+		return hk[]
+	else
+		return 0
+	end
+end
+
+function regclosekey(key::HKEY)
+	if key == ROOT_HKEY.HKEY_CLASSES_ROOT ||
+	  key == ROOT_HKEY.HKEY_CURRENT_CONFIG ||
+	  key == ROOT_HKEY.HKEY_CURRENT_USER ||
+	  key == ROOT_HKEY.HKEY_LOCAL_MACHINE ||
+	  key == ROOT_HKEY.HKEY_USERS ||
+	  key == ROOT_HKEY.HKEY_PERFORMANCE_DATA ||
+	  key == ROOT_HKEY.HKEY_PERFORMANCE_TEXT ||
+	  key == ROOT_HKEY.HKEY_PERFORMANCE_NLSTEXT ||
+	  key == ROOT_HKEY.HKEY_CURRENT_CONFIG ||
+	  key == ROOT_HKEY.HKEY_DYN_DATA
+		return 0
+	end
+	
+	ret = ccall((:RegCloseKey, advapi), stdcall,
+	  LONG,
+	  (HKEY, ),
+	  key )
+	
+	return ret
+end
+
+function any2bytes(x)
+	local sz = sizeof(x)
+	local ba = Vector{UInt8}(sz)
+	local src_ptr = convert(Ptr{UInt8}, pointer_from_objref(x))
+	unsafe_copy!(pointer(ba), src_ptr, sz)
+	
+	return ba
+end
+
+#return byte-array with NULL-termination
+function string2wchar(str::AbstractString)
```

Comment:
You can use `reinterpret(UInt8, Base.cconvert(Cwstring, str))` (or the implicit `Cwstring` conversion via `ccall`).